### DISCUSSION TITLE
gulp vs. rollup vs. webpack Which package is suitable for generating bundles

### DISCUSSION THREAD

**Discussion by yureshwar** [Digital-Assistant/Digital_Assistant_Client#330] [Ideas]
We need to finalize which package is good for bundling the codebase

https://npmcompare.com/compare/grunt,gulp,rollup,webpack

Comparing grunt vs. gulp vs. rollup vs. webpack

**Comment by kcramakrishna** [comment]
npm trends: Webpack is the most popular by far: https://npmcompare.com/compare/grunt,gulp,rollup,webpack 

https://ritza.co/articles/gen-articles/webpack-vs-babel-vs-rollup-vs-gulp-vs-parcel-vs-vite/ 

I think we should still use gulp for minification, unit-testing, linting etc.

Snowpack: Some benefits of using Snowpack in development: https://blog.sessionstack.com/how-javascript-works-a-guide-to-build-tools-exploring-webpack-parcel-rollup-es-build-and-2089bcf0ddb4 

Browserify: https://survivejs.com/webpack/appendices/comparison/

Rollup can be used with webpack: https://www.npmjs.com/package/rollup-loader
We can use rollup in case we want external libraries to use out code. i.e. some Robot software can import our module and get only specific functionality.

Concept of Tree-shaking to use only required features in imports: https://www.codefeetime.com/post/tree-shaking-a-react-component-library-in-rollup/ 

All in alls very cool stuff. Looks like webpack is most popular. 
Does it make sense using rollup-loader  ?

**Comment by lakshmanveti** [comment]
My recommendation is to go with "Webpack", not just because it is widely used, but the way it optimizes the bundling.

At its core Webpack allows us to use javascript modules within the project by taking multiple files and assets and combining them into one big file as shown below in this image from the new docs for Webpack 2.

![image](https://user-images.githubusercontent.com/3147681/178697132-f21d35da-824d-4a52-a115-410868896be9.png)

Besides it's efficiency to bundle the code, it is well maintained and vastly used.

**Comment by kcramakrishna** [comment]
Came across this alternative for webpack. Apparently the author of webpack released this new build tool. Apparently this is a drop-in replacement. i.e. replacement effort should be trivial. We should seriously consider adopting it.

https://devclass.com/2022/10/25/webpack-founder-debuts-rust-based-turbopack-that-is-700x-faster/

**Comment by kcramakrishna** [comment]
How to manage multiple node versions in the same system for different node projects
https://computingforgeeks.com/how-to-run-multiple-versions-of-node-js-on-linux/
https://www.baeldung.com/linux/multiple-node-js-versions

[kc]$ sudo pacman -Sy nvm
[kc]$ echo 'source /usr/share/nvm/init-nvm.sh' >> ~/.bashrc
[kc]$ echo 'source /usr/share/nvm/init-nvm.sh' >> ~/.zshrc
[kc]$ source ~/.bashrc
[kc]$ nvm install --lts
Installing latest LTS version.
Downloading and installing node v18.16.0...
Downloading https://nodejs.org/dist/v18.16.0/node-v18.16.0-linux-x64.tar.xz...
####################################################################################################### 100.0%
Computing checksum with sha256sum
Checksums matched!
Now using node v18.16.0 (npm v9.5.1)
Creating default alias: default -> lts/* (-> v18.16.0)
[kc]$ node -v
v18.16.0
[kc]$ nvm -v
0.39.3
[kc]$ which node
/home/kc/.nvm/versions/node/v18.16.0/bin/node

**Comment by kcramakrishna** [comment]
Improving performance of webpack in development.
https://webpack.js.org/guides/build-performance/
https://webpack.js.org/configuration/cache/

Issue created: #428
